### ISSUE TITLE
Bad PCC on GPT-OSS 120B TP8-DP4-EP32 Galaxy benchmark

### ISSUE THREAD

**Issue by odjuricicTT** [tenstorrent/tt-xla#4206]
## Description
The new GPT-OSS 120B TP8-DP4-EP32 benchmark on Galaxy (4x8 mesh, batch size 128) produces bad PCC accuracy results that need investigation.

## Context
- Introduced in #4205
- Test: `test_gpt_oss_120b_tp_dp_galaxy_batch_size_128`
- Configuration: TP=8, DP=4, EP=32
- Perf benchmark run: https://github.com/tenstorrent/tt-xla/actions/runs/24247948813

**Comment by odjuricicTT** [comment]
## PCC Investigation Plan

Starting investigation of bad PCC on prefill logits for the GPT-OSS 120B TP8-DP4-EP32 Galaxy benchmark (`--num-layers 1`).

### Analysis of TP-DP shard spec vs TP-only shard spec

Comparing `_moe_throughput_galaxy_shard_spec_fn` (TP-DP, used by this test) vs `ModelLoader.load_shard_spec()` (TP-only, used by `test_gpt_oss_120b_tp_galaxy_batch_size_64`):

| Parameter | TP-only (loader) | TP-DP (custom fn) |
|-----------|-------------------|-------------------|
| `batch_axis` | `"batch"` | `None` (replicated) |
| `q/k/v_proj.bias` | `("model",)` sharded | **Missing entirely** |
| `o_proj.bias` | `("batch",)` sharded | **Missing entirely** |
| `lm_head.weight` | `("model", "batch")` | `(None, None)` replicated |
| `sinks` | `(None,)` replicated | `("model",)` sharded |
| Expert sharding | `("model", "batch", None)` | `(("model", "batch"), None, None)` |

### Key hypotheses
1. **Missing attention bias shardings** in TP-DP shard spec — compiler propagation may assign incorrect shardings
2. **Stale KV cache** — accuracy benchmark reuses cache from warmup+perf runs
3. **MoE expert routing sensitivity** — numerical differences flipping top-4 expert selection

### Phase 0: Characterizing the failure
Added diagnostics to capture: actual PCC value, NaN/Inf checks, shape/dtype/range, argmax token agreement, and per-position PCC breakdown. Test running now.

**Comment by odjuricicTT** [comment]
## Phase 0: Failure Characterization

Test: `test_gpt_oss_120b_tp_dp_galaxy_batch_size_128 --num-layers 1`

### Results

- **Overall PCC: 0.928298** (required: 0.94)
- **Logit shapes**: `[17, 201088]` (17 seq positions, 201K vocab) — consistent between CPU and device
- **Value ranges**: CPU `[-20.5, 18.9]` vs DEV `[-22.0, 19.3]` — similar
- **No NaN/Inf** in either output
- **Argmax agreement: only 47%** (9/17 positions predict different top tokens!)

### Per-Position PCC Breakdown

```
pos  0: PCC=0.9799
pos  1: PCC=0.8430  ← argmax mismatch
pos  2: PCC=0.8969
pos  3: PCC=0.9922  (best)
pos  4: PCC=0.8703  ← argmax mismatch
pos  5: PCC=0.9123  ← argmax mismatch
pos  6: PCC=0.8128  (worst) ← argmax mismatch
pos  7: PCC=0.9599  ← argmax mismatch
pos  8: PCC=0.9509  ← argmax mismatch
pos  9: PCC=0.9626
pos 10: PCC=0.9903  ← argmax mismatch
pos 11: PCC=0.9421
pos 12: PCC=0.9228
pos 13: PCC=0.9541  ← argmax mismatch
pos 14: PCC=0.9695
pos 15: PCC=0.9119  ← argmax mismatch
pos 16: PCC=0.9552
```

Mean per-position PCC: 0.931, min: 0.813, max: 0.992

### Analysis

The 47% argmax disagreement with similar value ranges suggests a structural issue rather than simple quantization noise. PCC degradation is spread across positions (not isolated to a few).

**Top hypothesis**: The accuracy benchmark reuses the KV cache from the warmup+perf runs (which has stale data from 111 decode steps). The CPU baseline uses a fresh zeroed cache. During prefill, positions 0-16 are written, but positions 17-127 contain stale non-zero data. If the attention mask doesn't fully exclude these stale positions, the outputs diverge.

**Testing now**: Running with a fresh KV cache for the accuracy benchmark.

**Comment by odjuricicTT** [comment]
## Phase 1 Progress

### 1b. Fresh KV Cache test — **ELIMINATED**

Used `past_key_values=None` in accuracy benchmark to create a fresh zeroed StaticCache (matching the CPU baseline). Result:

**PCC = 0.928298** — exactly identical to the stale cache run. The attention mask correctly handles stale cache data. Stale KV cache is NOT the cause.

### 1a. Missing Bias Shardings — **TESTING NOW**

The TP-DP shard spec (`_moe_throughput_galaxy_shard_spec_fn`) omits all attention biases, while the TP-only shard spec (`ModelLoader.load_shard_spec`) explicitly shards them:

```python
# Missing from TP-DP, present in TP-only:
shard_specs[layer.self_attn.q_proj.bias] = ("model",)
shard_specs[layer.self_attn.k_proj.bias] = ("model",)
shard_specs[layer.self_attn.v_proj.bias] = ("model",)
shard_specs[layer.self_attn.o_proj.bias] = (batch_axis,)  # (None,)
```

Added these biases to the TP-DP shard spec. Running test now.

**Comment by odjuricicTT** [comment]
## Critical Finding: TP-only vs TP-DP Comparison

### TP-only 120B (`test_gpt_oss_120b_tp_galaxy_batch_size_64 --num-layers 1`)
- **PCC = 0.999968** ✅
- All positions: PCC > 0.9995
- 100% argmax agreement
- Uses `ModelLoader.load_shard_spec()`

### TP-DP 120B (`test_gpt_oss_120b_tp_dp_galaxy_batch_size_128 --num-layers 1`)
- **PCC = 0.928298** ❌
- Positions range: 0.81 — 0.99
- 47% argmax agreement
- Uses `_moe_throughput_galaxy_shard_spec_fn()`

### Conclusion
The issue is **definitively in the TP-DP shard spec**, not in BFP8 quantization or hardware computation. The TP-only test achieves near-perfect PCC with the same model on the same hardware.

### Key shard spec differences to investigate
| Parameter | TP-only (PCC=0.9999) | TP-DP (PCC=0.928) |
|-----------|----------------------|-------------------|
| Expert sharding (gate_up_proj) | `("model", "batch", None)` | `(("model", "batch"), None, None)` |
| Expert sharding (down_proj) | `("model", None, "batch")` | `(("model", "batch"), None, None)` |
| lm_head.weight | `("model", "batch")` | `(None, None)` |
| sinks | `(None,)` | `("model",)` |
| batch_axis on weights | `"batch"` (sharded) | `None` (replicated) |

### Also eliminated
- ❌ Stale KV cache: PCC identical with fresh cache
- ❌ Missing bias shardings: PCC identical when adding explicit bias shardings

### Next step
Systematically change the TP-DP shard spec to match the TP-only spec, one parameter group at a time, to identify which specific difference causes the PCC drop.

**Comment by odjuricicTT** [comment]
## Root Cause Identified: Expert Sharding Pattern

### The problem

The expert sharding pattern in `_moe_throughput_galaxy_shard_spec_fn` uses:
```python
shard_specs[layer.mlp.experts.gate_up_proj] = (("model", "batch"), None, None)
shard_specs[layer.mlp.experts.gate_up_proj_bias] = (("model", "batch"), None)
shard_specs[layer.mlp.experts.down_proj] = (("model", "batch"), None, None)
shard_specs[layer.mlp.experts.down_proj_bias] = (("model", "batch"), None)
```

This combined-axis sharding `(("model", "batch"), ...)` distributes 128 experts across all 32 devices (4 per device) but leaves the hidden/expert dimensions fully replicated.

### The fix

Using the TP-only sharding pattern (from `ModelLoader.load_shard_spec`):
```python
shard_specs[layer.mlp.experts.gate_up_proj] = ("model", "batch", None)
shard_specs[layer.mlp.experts.gate_up_proj_bias] = ("model", None)
shard_specs[layer.mlp.experts.down_proj] = ("model", None, "batch")
shard_specs[layer.mlp.experts.down_proj_bias] = ("model", "batch")
```

This shards experts across "model" axis (8-way) AND shards the hidden dimension across "batch" axis (4-way) separately.

### Results comparison

| Metric | Original TP-DP EP sharding | TP-only EP sharding |
|--------|---------------------------|-------------------|
| **Overall PCC** | **0.928298** ❌ | **0.999816** ✅ |
| Argmax agreement | 47% | 94% |
| Worst per-pos PCC | 0.812 | 0.998 |

### Eliminated candidates
- ❌ Stale KV cache (PCC=0.928298 with fresh cache)
- ❌ Missing bias shardings (PCC=0.928298 with explicit biases)
- ❌ Sinks sharding `("model",)` vs `(None,)` (PCC=0.928298 with either)
- ✅ **Expert sharding pattern is the root cause**

### Analysis

The combined-axis sharding `(("model", "batch"), None, None)` appears to produce incorrect all-to-all communication patterns for the MoE expert weights. The separate per-axis sharding `("model", "batch", None)` distributes experts and hidden dimensions independently, resulting in correct all-reduce/all-gather operations.

This may be a Shardy/XLA SPMD issue with combined-axis sharding, or it may indicate that the combined-axis pattern doesn't correctly represent the intended expert parallelism layout.

**Comment by odjuricicTT** [comment]
## Full Model Test Result: OOM

The fix (using separate per-axis expert sharding `("model", "batch", None)`) passes PCC with 1 layer but **OOMs on the full 36-layer model**:

```
TT_FATAL: Out of Memory: Not enough space to allocate 1647312896 B DRAM buffer across 12 banks,
where each bank needs to store 137277440 B, but bank size is 1071821792 B
(allocated: 892403968 B, free: 179417824 B, largest free block: 96983264 B)
```

This is expected — the TP-only sharding `("model", "batch", None)` shards the hidden dimension across the "batch" axis (4-way), which changes the per-device memory footprint. With 36 layers of MoE experts, the sharded weights + activations exceed device DRAM.

### The original combined-axis sharding `(("model", "batch"), None, None)` was designed to fit in memory by:
- Sharding 128 experts across all 32 devices (4 per device)
- Keeping hidden/expert dimensions fully replicated per device
- This avoids all-reduce on hidden dim but needs less total communication

### Root cause confirmed
The combined-axis sharding pattern produces incorrect PCC. A different expert sharding strategy that both fits in memory AND produces correct results is needed.

### Possible approaches:
1. Investigate why `(("model", "batch"), None, None)` produces wrong results — may be a Shardy/XLA SPMD compiler bug with combined-axis sharding
2. Find a different sharding that fits in memory (e.g., shard expert_dim instead of hidden_dim)
3. Use mixed precision to reduce per-expert memory

**Comment by mvasiljevicTT** [comment]
There are two separate problems here:
- `scatter op` in monkey patch for `GptOssTopKRouter` is wrongly lowered, as explained in [the discussion](https://github.com/tenstorrent/tt-mlir/discussions/7334#discussioncomment-16480883), this will not be relevant once MoE is rewritten, quick fix in the meantime is to use `values*one_hot(indices)` instead of `zeros.scatter_(indices,values)`.  With this fix, all devices will work the same, each having one good and 31 bad outputs, instead of all 127 being bad. 
- `sdpa_decode` device op assumes attention mask is broadcasted on all users, which isn't currently done in tt-mlir. With the change in tt-mlir now both gpt-oss and llama on galaxy have good outputs on all users. 

Along this I noticed few new issues:
- `sdpa composite` is not matching in either llama or gpt-oss
- if `sdpa` fails validation during fusing but after the workaround, the workaround will stay, leaving the graph in a different state than before matching
- `sdpa_decode` device op doesn't check shape of the mask, and reads garbage data when smaller tensor is given. If that is block of zeros, it won't affect the model as much as there are larger values, which might be the reason why some models worked without this change.

Will add PRs and issues soon.
